{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import gzip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"/home/marc/inters/output_test2/05_Annotation/Nirvana/snvs/Annotation.json.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of positions: 6557\n",
      "number of genes: 313\n"
     ]
    }
   ],
   "source": [
    "\n",
    "header = ''\n",
    "positions = []\n",
    "genes = []\n",
    "is_header_line = True\n",
    "is_position_line = False\n",
    "is_gene_line = False\n",
    "gene_section_line = '],\"genes\":['\n",
    "end_line = ']}'\n",
    "\n",
    "with gzip.open(file, 'rt') as f:\n",
    "    position_count = 0\n",
    "    gene_count = 0\n",
    "    for line in f:\n",
    "        trim_line = line.strip()\n",
    "        if is_header_line:\n",
    "            ## only keep the \"header\" field content from the line\n",
    "            header = trim_line[10:-14]\n",
    "            is_header_line = False\n",
    "            is_position_line = True\n",
    "            continue\n",
    "        if trim_line == gene_section_line:\n",
    "            is_gene_line = True\n",
    "            is_position_line = False\n",
    "            continue\n",
    "        elif trim_line == end_line:\n",
    "            break\n",
    "        else:\n",
    "            if is_position_line:\n",
    "                ## remove the trailing ',' if there is\n",
    "                positions.append(trim_line.rstrip(','))\n",
    "                position_count += 1\n",
    "            if is_gene_line:\n",
    "                ## remove the trailing ',' if there is\n",
    "                genes.append(trim_line.rstrip(','))\n",
    "                gene_count += 1\n",
    "\n",
    "print ('number of positions:', position_count)\n",
    "print ('number of genes:', gene_count)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "mylist = []\n",
    "for gene in genes:\n",
    "    position_dict = json.loads(gene)\n",
    "    mylist.append(position_dict)\n",
    "genedf = pd.DataFrame(mylist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gnomAD</th>\n",
       "      <th>omim</th>\n",
       "      <th>clingenDosageSensitivityMap</th>\n",
       "      <th>clingenGeneValidity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC099805.1</td>\n",
       "      <td>{'pLi': 0.104, 'pRec': 0.604, 'pNull': 0.292, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC100868.1</td>\n",
       "      <td>{'pLi': 5.68e-12, 'pRec': 0.114, 'pNull': 0.88...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC105001.2</td>\n",
       "      <td>{'pLi': 3.48e-07, 'pRec': 0.571, 'pNull': 0.42...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC110275.1</td>\n",
       "      <td>{'pLi': 0.846, 'pRec': 0.154, 'pNull': 0.00010...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAM18</td>\n",
       "      <td>{'pLi': 6.88e-35, 'pRec': 3.53e-07, 'pNull': 1...</td>\n",
       "      <td>[{'mimNumber': 619495, 'geneName': 'ADAM metal...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>ZNF696</td>\n",
       "      <td>{'pLi': 0.207, 'pRec': 0.755, 'pNull': 0.038, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>ZNF703</td>\n",
       "      <td>{'pLi': 0.356, 'pRec': 0.632, 'pNull': 0.0113,...</td>\n",
       "      <td>[{'mimNumber': 617045, 'geneName': 'Zinc finge...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ZNF704</td>\n",
       "      <td>{'pLi': 0.995, 'pRec': 0.00468, 'pNull': 8.22e...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>ZNF706</td>\n",
       "      <td>{'pLi': 0.702, 'pRec': 0.283, 'pNull': 0.0146,...</td>\n",
       "      <td>[{'mimNumber': 619526, 'geneName': 'Zinc finge...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>ZNF707</td>\n",
       "      <td>{'pLi': 0.00131, 'pRec': 0.87, 'pNull': 0.129,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>313 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           name                                             gnomAD  \\\n",
       "0    AC099805.1  {'pLi': 0.104, 'pRec': 0.604, 'pNull': 0.292, ...   \n",
       "1    AC100868.1  {'pLi': 5.68e-12, 'pRec': 0.114, 'pNull': 0.88...   \n",
       "2    AC105001.2  {'pLi': 3.48e-07, 'pRec': 0.571, 'pNull': 0.42...   \n",
       "3    AC110275.1  {'pLi': 0.846, 'pRec': 0.154, 'pNull': 0.00010...   \n",
       "4        ADAM18  {'pLi': 6.88e-35, 'pRec': 3.53e-07, 'pNull': 1...   \n",
       "..          ...                                                ...   \n",
       "308      ZNF696  {'pLi': 0.207, 'pRec': 0.755, 'pNull': 0.038, ...   \n",
       "309      ZNF703  {'pLi': 0.356, 'pRec': 0.632, 'pNull': 0.0113,...   \n",
       "310      ZNF704  {'pLi': 0.995, 'pRec': 0.00468, 'pNull': 8.22e...   \n",
       "311      ZNF706  {'pLi': 0.702, 'pRec': 0.283, 'pNull': 0.0146,...   \n",
       "312      ZNF707  {'pLi': 0.00131, 'pRec': 0.87, 'pNull': 0.129,...   \n",
       "\n",
       "                                                  omim  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4    [{'mimNumber': 619495, 'geneName': 'ADAM metal...   \n",
       "..                                                 ...   \n",
       "308                                                NaN   \n",
       "309  [{'mimNumber': 617045, 'geneName': 'Zinc finge...   \n",
       "310                                                NaN   \n",
       "311  [{'mimNumber': 619526, 'geneName': 'Zinc finge...   \n",
       "312                                                NaN   \n",
       "\n",
       "    clingenDosageSensitivityMap clingenGeneValidity  \n",
       "0                           NaN                 NaN  \n",
       "1                           NaN                 NaN  \n",
       "2                           NaN                 NaN  \n",
       "3                           NaN                 NaN  \n",
       "4                           NaN                 NaN  \n",
       "..                          ...                 ...  \n",
       "308                         NaN                 NaN  \n",
       "309                         NaN                 NaN  \n",
       "310                         NaN                 NaN  \n",
       "311                         NaN                 NaN  \n",
       "312                         NaN                 NaN  \n",
       "\n",
       "[313 rows x 5 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "mylist = []\n",
    "for position in positions:\n",
    "    position_dict = json.loads(position)\n",
    "    mylist.append(position_dict)\n",
    "df = pd.DataFrame(mylist)\n",
    "variants_df = df['variants'].apply(pd.Series)\n",
    "variants_df2 = variants_df[0].apply(pd.Series)\n",
    "final_df = pd.concat([df, variants_df2], axis=1).drop(['variants', 'samples'], axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'transcript': 'ENST00000320901.3',\n",
       "  'source': 'Ensembl',\n",
       "  'bioType': 'protein_coding',\n",
       "  'geneId': 'ENSG00000176269',\n",
       "  'hgnc': 'OR4F21',\n",
       "  'consequence': ['upstream_gene_variant'],\n",
       "  'isCanonical': True,\n",
       "  'proteinId': 'ENSP00000318878.3'},\n",
       " {'transcript': 'NM_001005504.1',\n",
       "  'source': 'RefSeq',\n",
       "  'bioType': 'protein_coding',\n",
       "  'geneId': '441308',\n",
       "  'hgnc': 'OR4F21',\n",
       "  'consequence': ['upstream_gene_variant'],\n",
       "  'isCanonical': True,\n",
       "  'proteinId': 'NP_001005504.1'}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df['transcripts'][30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = pd.read_csv('/home/marc/inters/output_test2/05_Annotation/Nirvana/snvs/Annotation.tsv', delimiter='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>refAllele</th>\n",
       "      <th>altAlleles</th>\n",
       "      <th>quality</th>\n",
       "      <th>filters</th>\n",
       "      <th>fisherStrandBias</th>\n",
       "      <th>mappingQuality</th>\n",
       "      <th>cytogeneticBand</th>\n",
       "      <th>vid</th>\n",
       "      <th>...</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>gnomad</th>\n",
       "      <th>topmed</th>\n",
       "      <th>globalAllele</th>\n",
       "      <th>regulatoryRegions</th>\n",
       "      <th>inLowComplexityRegion</th>\n",
       "      <th>oneKg</th>\n",
       "      <th>revel</th>\n",
       "      <th>clinvar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>116079</td>\n",
       "      <td>G</td>\n",
       "      <td>['A']</td>\n",
       "      <td>301.90</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>7.799</td>\n",
       "      <td>60.00</td>\n",
       "      <td>8p23.3</td>\n",
       "      <td>8-116079-G-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>116701</td>\n",
       "      <td>A</td>\n",
       "      <td>['G']</td>\n",
       "      <td>107.29</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>8p23.3</td>\n",
       "      <td>8-116701-A-G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>116895</td>\n",
       "      <td>A</td>\n",
       "      <td>['G']</td>\n",
       "      <td>161.28</td>\n",
       "      <td>['SOR3']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>8p23.3</td>\n",
       "      <td>8-116895-A-G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>160552</td>\n",
       "      <td>G</td>\n",
       "      <td>['A']</td>\n",
       "      <td>36.29</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>59.25</td>\n",
       "      <td>8p23.3</td>\n",
       "      <td>8-160552-G-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>160608</td>\n",
       "      <td>A</td>\n",
       "      <td>['C']</td>\n",
       "      <td>976.97</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>8p23.3</td>\n",
       "      <td>8-160608-A-C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6552</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146200492</td>\n",
       "      <td>G</td>\n",
       "      <td>['A']</td>\n",
       "      <td>1241.95</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8-146200492-G-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6553</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146203412</td>\n",
       "      <td>A</td>\n",
       "      <td>['G']</td>\n",
       "      <td>8533.95</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>3.114</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8-146203412-A-G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6554</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146220094</td>\n",
       "      <td>C</td>\n",
       "      <td>['T']</td>\n",
       "      <td>2706.95</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8-146220094-C-T</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6555</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146229040</td>\n",
       "      <td>T</td>\n",
       "      <td>['G']</td>\n",
       "      <td>616.95</td>\n",
       "      <td>['PASS']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8-146229040-T-G</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6556</th>\n",
       "      <td>chr8</td>\n",
       "      <td>146229161</td>\n",
       "      <td>T</td>\n",
       "      <td>['A']</td>\n",
       "      <td>845.08</td>\n",
       "      <td>['SOR3']</td>\n",
       "      <td>0.000</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8-146229161-T-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6557 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome   position refAllele altAlleles  quality   filters  \\\n",
       "0          chr8     116079         G      ['A']   301.90  ['PASS']   \n",
       "1          chr8     116701         A      ['G']   107.29  ['PASS']   \n",
       "2          chr8     116895         A      ['G']   161.28  ['SOR3']   \n",
       "3          chr8     160552         G      ['A']    36.29  ['PASS']   \n",
       "4          chr8     160608         A      ['C']   976.97  ['PASS']   \n",
       "...         ...        ...       ...        ...      ...       ...   \n",
       "6552       chr8  146200492         G      ['A']  1241.95  ['PASS']   \n",
       "6553       chr8  146203412         A      ['G']  8533.95  ['PASS']   \n",
       "6554       chr8  146220094         C      ['T']  2706.95  ['PASS']   \n",
       "6555       chr8  146229040         T      ['G']   616.95  ['PASS']   \n",
       "6556       chr8  146229161         T      ['A']   845.08  ['SOR3']   \n",
       "\n",
       "      fisherStrandBias  mappingQuality cytogeneticBand              vid  ...  \\\n",
       "0                7.799           60.00          8p23.3     8-116079-G-A  ...   \n",
       "1                0.000           60.00          8p23.3     8-116701-A-G  ...   \n",
       "2                0.000           60.00          8p23.3     8-116895-A-G  ...   \n",
       "3                0.000           59.25          8p23.3     8-160552-G-A  ...   \n",
       "4                0.000           60.00          8p23.3     8-160608-A-C  ...   \n",
       "...                ...             ...             ...              ...  ...   \n",
       "6552             0.000           60.00             NaN  8-146200492-G-A  ...   \n",
       "6553             3.114           60.00             NaN  8-146203412-A-G  ...   \n",
       "6554             0.000           60.00             NaN  8-146220094-C-T  ...   \n",
       "6555             0.000           60.00             NaN  8-146229040-T-G  ...   \n",
       "6556             0.000           60.00             NaN  8-146229161-T-A  ...   \n",
       "\n",
       "     transcripts  dbsnp  gnomad topmed globalAllele regulatoryRegions  \\\n",
       "0            NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "1            NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "2            NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "3            NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "4            NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "...          ...    ...     ...    ...          ...               ...   \n",
       "6552         NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "6553         NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "6554         NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "6555         NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "6556         NaN    NaN     NaN    NaN          NaN               NaN   \n",
       "\n",
       "     inLowComplexityRegion  oneKg  revel  clinvar  \n",
       "0                      NaN    NaN    NaN      NaN  \n",
       "1                      NaN    NaN    NaN      NaN  \n",
       "2                      NaN    NaN    NaN      NaN  \n",
       "3                      NaN    NaN    NaN      NaN  \n",
       "4                      NaN    NaN    NaN      NaN  \n",
       "...                    ...    ...    ...      ...  \n",
       "6552                   NaN    NaN    NaN      NaN  \n",
       "6553                   NaN    NaN    NaN      NaN  \n",
       "6554                   NaN    NaN    NaN      NaN  \n",
       "6555                   NaN    NaN    NaN      NaN  \n",
       "6556                   NaN    NaN    NaN      NaN  \n",
       "\n",
       "[6557 rows x 30 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Variants: 0\n",
      "Variants with Annotations: 0\n",
      "Variants without dbSNP Annotations: 0\n",
      "Percentage of Novel Variants: nan %\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/marc/.local/lib/python3.6/site-packages/ipykernel_launcher.py:26: RuntimeWarning: invalid value encountered in long_scalars\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Define the gene symbol you want to search for\n",
    "target_gene_symbol = 'P53'\n",
    "# final_df = pd.read_csv('/home/marc/inters/output_test2/05_Annotation/Nirvana/snvs/Annotation.tsv', delimiter='\\t')\n",
    "\n",
    "# Function to filter rows based on the target gene symbol\n",
    "def filter_by_gene_symbol(row):\n",
    "    if isinstance(row, list):\n",
    "        for d in row:\n",
    "            if isinstance(d, dict) and d.get('hgnc') == target_gene_symbol:\n",
    "                return True\n",
    "    return False\n",
    "\n",
    "# Apply the filter function to the DataFrame\n",
    "filtered_df = final_df[final_df['transcripts'].apply(filter_by_gene_symbol)]\n",
    "\n",
    "# Print the filtered DataFrame\n",
    "# print(filtered_df)\n",
    "\n",
    "def is_novel(row):\n",
    "    return pd.isna(row['dbsnp'])\n",
    "\n",
    "total_rows = len(filtered_df)\n",
    "total_variants = total_rows\n",
    "variants_with_annotations = total_rows - filtered_df['dbsnp'].isna().sum()\n",
    "novel_variants = total_variants - variants_with_annotations\n",
    "percentage_novel = (novel_variants / total_rows) * 100\n",
    "# Print the statistics\n",
    "print()\n",
    "print(f\"#--------------------{target_gene_symbol}--------------------#\")\n",
    "print(f\"Total Variants: {total_variants}\")\n",
    "print(f\"Variants with Annotations: {variants_with_annotations}\")\n",
    "print(f\"Variants without dbSNP Annotations: {novel_variants}\")\n",
    "print(f\"Percentage of Novel Variants: {percentage_novel} %\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "empty\n"
     ]
    }
   ],
   "source": [
    "if not filtered_df.empty:\n",
    "    pass\n",
    "else:\n",
    "    print(\"empty\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
