{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snakemake.io import *\n",
    "import snakemake\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "wildcards = Wildcards()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "wildcards.unit = \"ACC1-LIBR_L1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACC1-LIBR_L2'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wildcards.unit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_table_file='/home/marc/Desktop/output_new_wes_gatk/samples.tsv'\n",
    "SampleTable = pd.read_table(sample_table_file)\n",
    "\n",
    "\n",
    "files_R1s = list(SampleTable.iloc[:, 1])\n",
    "files_R2s = list(SampleTable.iloc[:, 16])\n",
    "samples = list(SampleTable.iloc[:, 3]) # sample full name\n",
    "units = list(SampleTable.iloc[:, 2])\n",
    "samples_IDs = list(SampleTable.iloc[:, 4])\n",
    "library_index = list(SampleTable.iloc[:, 5])\n",
    "\n",
    "units = (\n",
    "    pd.read_csv('/home/marc/Desktop/output_new_wes_gatk/samples.tsv', sep=\"\\t\", dtype={\"sample_id\": str, \"library_index\": str, \"lane\": str})\n",
    "    .set_index([\"sample_id\", \"unit\"], drop=False)\n",
    "    .sort_index()\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>file_name</th>\n",
       "      <th>unit</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>library_index</th>\n",
       "      <th>acc1</th>\n",
       "      <th>acc2</th>\n",
       "      <th>lane</th>\n",
       "      <th>R_pattern</th>\n",
       "      <th>R_sep</th>\n",
       "      <th>read_num</th>\n",
       "      <th>ext</th>\n",
       "      <th>tail</th>\n",
       "      <th>sample_number</th>\n",
       "      <th>matched_pattern</th>\n",
       "      <th>file2</th>\n",
       "      <th>PE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>father_ACC1-LIBR_L2</td>\n",
       "      <td>father_ACC1-LIBR_L2_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L2</td>\n",
       "      <td>father_ACC1-LIBR_L2</td>\n",
       "      <td>father</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L2</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>father_ACC1-LIBR_L2_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>father_ACC1-LIBR_L1</td>\n",
       "      <td>father_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>father_ACC1-LIBR_L1</td>\n",
       "      <td>father</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>father_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mother_ACC1-LIBR_L1</td>\n",
       "      <td>mother_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>mother_ACC1-LIBR_L1</td>\n",
       "      <td>mother</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>mother_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mother_ACC2-LIBR2_L1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1_1.fq.gz</td>\n",
       "      <td>ACC2-LIBR2_L1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1</td>\n",
       "      <td>mother</td>\n",
       "      <td>ACC2-LIBR2</td>\n",
       "      <td>ACC2</td>\n",
       "      <td>LIBR2</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>proband_ACC1-LIBR_L1</td>\n",
       "      <td>proband_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>proband_ACC1-LIBR_L1</td>\n",
       "      <td>proband</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>proband_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>proband_ACC2-LIBR2_L3</td>\n",
       "      <td>proband_ACC2-LIBR2_L3_1.fq.gz</td>\n",
       "      <td>ACC2-LIBR2_L3</td>\n",
       "      <td>proband_ACC2-LIBR2_L3</td>\n",
       "      <td>proband</td>\n",
       "      <td>ACC2-LIBR2</td>\n",
       "      <td>ACC2</td>\n",
       "      <td>LIBR2</td>\n",
       "      <td>L3</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>proband_ACC2-LIBR2_L3_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Unnamed: 0                      file_name           unit  \\\n",
       "0    father_ACC1-LIBR_L2    father_ACC1-LIBR_L2_1.fq.gz   ACC1-LIBR_L2   \n",
       "1    father_ACC1-LIBR_L1    father_ACC1-LIBR_L1_1.fq.gz   ACC1-LIBR_L1   \n",
       "2    mother_ACC1-LIBR_L1    mother_ACC1-LIBR_L1_1.fq.gz   ACC1-LIBR_L1   \n",
       "3   mother_ACC2-LIBR2_L1   mother_ACC2-LIBR2_L1_1.fq.gz  ACC2-LIBR2_L1   \n",
       "4   proband_ACC1-LIBR_L1   proband_ACC1-LIBR_L1_1.fq.gz   ACC1-LIBR_L1   \n",
       "5  proband_ACC2-LIBR2_L3  proband_ACC2-LIBR2_L3_1.fq.gz  ACC2-LIBR2_L3   \n",
       "\n",
       "             sample_name sample_id library_index  acc1   acc2 lane R_pattern  \\\n",
       "0    father_ACC1-LIBR_L2    father     ACC1-LIBR  ACC1   LIBR   L2        _1   \n",
       "1    father_ACC1-LIBR_L1    father     ACC1-LIBR  ACC1   LIBR   L1        _1   \n",
       "2    mother_ACC1-LIBR_L1    mother     ACC1-LIBR  ACC1   LIBR   L1        _1   \n",
       "3   mother_ACC2-LIBR2_L1    mother    ACC2-LIBR2  ACC2  LIBR2   L1        _1   \n",
       "4   proband_ACC1-LIBR_L1   proband     ACC1-LIBR  ACC1   LIBR   L1        _1   \n",
       "5  proband_ACC2-LIBR2_L3   proband    ACC2-LIBR2  ACC2  LIBR2   L3        _1   \n",
       "\n",
       "  R_sep  read_num    ext  tail  sample_number matched_pattern  \\\n",
       "0     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "1     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "2     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "3     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "4     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "5     _         1  fq.gz   NaN            NaN        Novagen1   \n",
       "\n",
       "                           file2    PE  \n",
       "0    father_ACC1-LIBR_L2_2.fq.gz  True  \n",
       "1    father_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "2    mother_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "3   mother_ACC2-LIBR2_L1_2.fq.gz  True  \n",
       "4   proband_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "5  proband_ACC2-LIBR2_L3_2.fq.gz  True  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SampleTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACC1-LIBR'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "units.loc[:, 'library_index'][units['unit'] == f\"ACC1-LIBR_L2\"].tolist()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACC1-LIBR'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(lambda wildcards: units.loc[:, 'library_index'][units['unit'] == f\"{wildcards.unit}\"].tolist()[0])(wildcards)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['father', 'father', 'mother', 'mother', 'proband', 'proband']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_final_output(wildcards):\n",
    "    final_output = []\n",
    "\n",
    "    for i in samples_IDs:\n",
    "        final_output.extend(expand(\n",
    "            \"01_QC/{sample}_{unit}_{R}_fastqc.zip\",\n",
    "            sample=i,\n",
    "            unit=units.loc[i, \"unit\"].tolist(),\n",
    "            R=[1,2]\n",
    "        ))\n",
    "        final_output.extend(expand(\n",
    "            \"02_alignment/QC/{sample}_{unit}.cov\",\n",
    "                sample = i,\n",
    "                unit=units.loc[i, \"unit\"].tolist()\n",
    "        ))\n",
    "        final_output.extend(expand(\n",
    "            \"03_bamPrep/QC/{sample}_{unit}_Qualimap\",\n",
    "                sample = i,\n",
    "                unit=units.loc[i, \"unit\"].tolist()\n",
    "        ))\n",
    "\n",
    "        final_output.extend(expand(\n",
    "            \"03_bamPrep/QC/{sample}_{unit}.pdf\",\n",
    "                sample = i,\n",
    "                unit=units.loc[i, \"unit\"].tolist()\n",
    "        ))\n",
    "    \n",
    "    final_output.extend(\n",
    "        [\n",
    "            \"04_calling/QC/bcftools.stats\",\n",
    "            \"04_calling/QC/bcftools_plots\",\n",
    "            \"04_calling/QC/bcftools_csq.vcf\",\n",
    "            \"04_calling/snvs/variants_genotyped.gvcf.gz\",\n",
    "            \"04_calling/indels/variants_genotyped.gvcf.gz\",\n",
    "            \"05_Annotation/Nirvana/Annotation.json.gz\",\n",
    "            \"05_Annotation/ANNOVAR\"\n",
    "        ]\n",
    "    )\n",
    "    final_output.extend(expand(\n",
    "        \"04_calling/QC/{sample}.eval.grp\",\n",
    "            sample = samples_IDs\n",
    "    ))\n",
    "\n",
    "    return final_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>file_name</th>\n",
       "      <th>unit</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>library_index</th>\n",
       "      <th>acc1</th>\n",
       "      <th>acc2</th>\n",
       "      <th>lane</th>\n",
       "      <th>R_pattern</th>\n",
       "      <th>R_sep</th>\n",
       "      <th>read_num</th>\n",
       "      <th>ext</th>\n",
       "      <th>tail</th>\n",
       "      <th>sample_number</th>\n",
       "      <th>matched_pattern</th>\n",
       "      <th>file2</th>\n",
       "      <th>PE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th>unit</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">father</th>\n",
       "      <th>ACC1-LIBR_L1</th>\n",
       "      <td>father_ACC1-LIBR_L1</td>\n",
       "      <td>father_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>father_ACC1-LIBR_L1</td>\n",
       "      <td>father</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>father_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC1-LIBR_L2</th>\n",
       "      <td>father_ACC1-LIBR_L2</td>\n",
       "      <td>father_ACC1-LIBR_L2_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L2</td>\n",
       "      <td>father_ACC1-LIBR_L2</td>\n",
       "      <td>father</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L2</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>father_ACC1-LIBR_L2_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">mother</th>\n",
       "      <th>ACC1-LIBR_L1</th>\n",
       "      <td>mother_ACC1-LIBR_L1</td>\n",
       "      <td>mother_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>mother_ACC1-LIBR_L1</td>\n",
       "      <td>mother</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>mother_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC2-LIBR2_L1</th>\n",
       "      <td>mother_ACC2-LIBR2_L1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1_1.fq.gz</td>\n",
       "      <td>ACC2-LIBR2_L1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1</td>\n",
       "      <td>mother</td>\n",
       "      <td>ACC2-LIBR2</td>\n",
       "      <td>ACC2</td>\n",
       "      <td>LIBR2</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>mother_ACC2-LIBR2_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">proband</th>\n",
       "      <th>ACC1-LIBR_L1</th>\n",
       "      <td>proband_ACC1-LIBR_L1</td>\n",
       "      <td>proband_ACC1-LIBR_L1_1.fq.gz</td>\n",
       "      <td>ACC1-LIBR_L1</td>\n",
       "      <td>proband_ACC1-LIBR_L1</td>\n",
       "      <td>proband</td>\n",
       "      <td>ACC1-LIBR</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>LIBR</td>\n",
       "      <td>L1</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>proband_ACC1-LIBR_L1_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACC2-LIBR2_L3</th>\n",
       "      <td>proband_ACC2-LIBR2_L3</td>\n",
       "      <td>proband_ACC2-LIBR2_L3_1.fq.gz</td>\n",
       "      <td>ACC2-LIBR2_L3</td>\n",
       "      <td>proband_ACC2-LIBR2_L3</td>\n",
       "      <td>proband</td>\n",
       "      <td>ACC2-LIBR2</td>\n",
       "      <td>ACC2</td>\n",
       "      <td>LIBR2</td>\n",
       "      <td>L3</td>\n",
       "      <td>_1</td>\n",
       "      <td>_</td>\n",
       "      <td>1</td>\n",
       "      <td>fq.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Novagen1</td>\n",
       "      <td>proband_ACC2-LIBR2_L3_2.fq.gz</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Unnamed: 0                      file_name  \\\n",
       "sample_id unit                                                                  \n",
       "father    ACC1-LIBR_L1     father_ACC1-LIBR_L1    father_ACC1-LIBR_L1_1.fq.gz   \n",
       "          ACC1-LIBR_L2     father_ACC1-LIBR_L2    father_ACC1-LIBR_L2_1.fq.gz   \n",
       "mother    ACC1-LIBR_L1     mother_ACC1-LIBR_L1    mother_ACC1-LIBR_L1_1.fq.gz   \n",
       "          ACC2-LIBR2_L1   mother_ACC2-LIBR2_L1   mother_ACC2-LIBR2_L1_1.fq.gz   \n",
       "proband   ACC1-LIBR_L1    proband_ACC1-LIBR_L1   proband_ACC1-LIBR_L1_1.fq.gz   \n",
       "          ACC2-LIBR2_L3  proband_ACC2-LIBR2_L3  proband_ACC2-LIBR2_L3_1.fq.gz   \n",
       "\n",
       "                                  unit            sample_name sample_id  \\\n",
       "sample_id unit                                                            \n",
       "father    ACC1-LIBR_L1    ACC1-LIBR_L1    father_ACC1-LIBR_L1    father   \n",
       "          ACC1-LIBR_L2    ACC1-LIBR_L2    father_ACC1-LIBR_L2    father   \n",
       "mother    ACC1-LIBR_L1    ACC1-LIBR_L1    mother_ACC1-LIBR_L1    mother   \n",
       "          ACC2-LIBR2_L1  ACC2-LIBR2_L1   mother_ACC2-LIBR2_L1    mother   \n",
       "proband   ACC1-LIBR_L1    ACC1-LIBR_L1   proband_ACC1-LIBR_L1   proband   \n",
       "          ACC2-LIBR2_L3  ACC2-LIBR2_L3  proband_ACC2-LIBR2_L3   proband   \n",
       "\n",
       "                        library_index  acc1   acc2 lane R_pattern R_sep  \\\n",
       "sample_id unit                                                            \n",
       "father    ACC1-LIBR_L1      ACC1-LIBR  ACC1   LIBR   L1        _1     _   \n",
       "          ACC1-LIBR_L2      ACC1-LIBR  ACC1   LIBR   L2        _1     _   \n",
       "mother    ACC1-LIBR_L1      ACC1-LIBR  ACC1   LIBR   L1        _1     _   \n",
       "          ACC2-LIBR2_L1    ACC2-LIBR2  ACC2  LIBR2   L1        _1     _   \n",
       "proband   ACC1-LIBR_L1      ACC1-LIBR  ACC1   LIBR   L1        _1     _   \n",
       "          ACC2-LIBR2_L3    ACC2-LIBR2  ACC2  LIBR2   L3        _1     _   \n",
       "\n",
       "                         read_num    ext  tail  sample_number matched_pattern  \\\n",
       "sample_id unit                                                                  \n",
       "father    ACC1-LIBR_L1          1  fq.gz   NaN            NaN        Novagen1   \n",
       "          ACC1-LIBR_L2          1  fq.gz   NaN            NaN        Novagen1   \n",
       "mother    ACC1-LIBR_L1          1  fq.gz   NaN            NaN        Novagen1   \n",
       "          ACC2-LIBR2_L1         1  fq.gz   NaN            NaN        Novagen1   \n",
       "proband   ACC1-LIBR_L1          1  fq.gz   NaN            NaN        Novagen1   \n",
       "          ACC2-LIBR2_L3         1  fq.gz   NaN            NaN        Novagen1   \n",
       "\n",
       "                                                 file2    PE  \n",
       "sample_id unit                                                \n",
       "father    ACC1-LIBR_L1     father_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "          ACC1-LIBR_L2     father_ACC1-LIBR_L2_2.fq.gz  True  \n",
       "mother    ACC1-LIBR_L1     mother_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "          ACC2-LIBR2_L1   mother_ACC2-LIBR2_L1_2.fq.gz  True  \n",
       "proband   ACC1-LIBR_L1    proband_ACC1-LIBR_L1_2.fq.gz  True  \n",
       "          ACC2-LIBR2_L3  proband_ACC2-LIBR2_L3_2.fq.gz  True  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "units"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ACC1-LIBR_L1', 'ACC1-LIBR_L2']\n",
      "['ACC1-LIBR_L1', 'ACC1-LIBR_L2']\n",
      "['ACC1-LIBR_L1', 'ACC2-LIBR2_L1']\n",
      "['ACC1-LIBR_L1', 'ACC2-LIBR2_L1']\n",
      "['ACC1-LIBR_L1', 'ACC2-LIBR2_L3']\n",
      "['ACC1-LIBR_L1', 'ACC2-LIBR2_L3']\n"
     ]
    }
   ],
   "source": [
    "for i in samples_IDs:\n",
    "    print(units.loc[i, \"unit\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['01_QC/father_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L2_1_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L2_2_fastqc.zip',\n",
       " '02_alignment/QC/father_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/father_ACC1-LIBR_L2.cov',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L2_Qualimap',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L2.pdf',\n",
       " '01_QC/father_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L2_1_fastqc.zip',\n",
       " '01_QC/father_ACC1-LIBR_L2_2_fastqc.zip',\n",
       " '02_alignment/QC/father_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/father_ACC1-LIBR_L2.cov',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L2_Qualimap',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/father_ACC1-LIBR_L2.pdf',\n",
       " '01_QC/mother_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/mother_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/mother_ACC2-LIBR2_L1_1_fastqc.zip',\n",
       " '01_QC/mother_ACC2-LIBR2_L1_2_fastqc.zip',\n",
       " '02_alignment/QC/mother_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/mother_ACC2-LIBR2_L1.cov',\n",
       " '03_bamPrep/QC/mother_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/mother_ACC2-LIBR2_L1_Qualimap',\n",
       " '03_bamPrep/QC/mother_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/mother_ACC2-LIBR2_L1.pdf',\n",
       " '01_QC/mother_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/mother_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/mother_ACC2-LIBR2_L1_1_fastqc.zip',\n",
       " '01_QC/mother_ACC2-LIBR2_L1_2_fastqc.zip',\n",
       " '02_alignment/QC/mother_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/mother_ACC2-LIBR2_L1.cov',\n",
       " '03_bamPrep/QC/mother_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/mother_ACC2-LIBR2_L1_Qualimap',\n",
       " '03_bamPrep/QC/mother_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/mother_ACC2-LIBR2_L1.pdf',\n",
       " '01_QC/proband_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/proband_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/proband_ACC2-LIBR2_L3_1_fastqc.zip',\n",
       " '01_QC/proband_ACC2-LIBR2_L3_2_fastqc.zip',\n",
       " '02_alignment/QC/proband_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/proband_ACC2-LIBR2_L3.cov',\n",
       " '03_bamPrep/QC/proband_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/proband_ACC2-LIBR2_L3_Qualimap',\n",
       " '03_bamPrep/QC/proband_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/proband_ACC2-LIBR2_L3.pdf',\n",
       " '01_QC/proband_ACC1-LIBR_L1_1_fastqc.zip',\n",
       " '01_QC/proband_ACC1-LIBR_L1_2_fastqc.zip',\n",
       " '01_QC/proband_ACC2-LIBR2_L3_1_fastqc.zip',\n",
       " '01_QC/proband_ACC2-LIBR2_L3_2_fastqc.zip',\n",
       " '02_alignment/QC/proband_ACC1-LIBR_L1.cov',\n",
       " '02_alignment/QC/proband_ACC2-LIBR2_L3.cov',\n",
       " '03_bamPrep/QC/proband_ACC1-LIBR_L1_Qualimap',\n",
       " '03_bamPrep/QC/proband_ACC2-LIBR2_L3_Qualimap',\n",
       " '03_bamPrep/QC/proband_ACC1-LIBR_L1.pdf',\n",
       " '03_bamPrep/QC/proband_ACC2-LIBR2_L3.pdf',\n",
       " '04_calling/QC/bcftools.stats',\n",
       " '04_calling/QC/bcftools_plots',\n",
       " '04_calling/QC/bcftools_csq.vcf',\n",
       " '04_calling/snvs/variants_genotyped.gvcf.gz',\n",
       " '04_calling/indels/variants_genotyped.gvcf.gz',\n",
       " '05_Annotation/Nirvana/Annotation.json.gz',\n",
       " '05_Annotation/ANNOVAR',\n",
       " '04_calling/QC/father.eval.grp',\n",
       " '04_calling/QC/father.eval.grp',\n",
       " '04_calling/QC/mother.eval.grp',\n",
       " '04_calling/QC/mother.eval.grp',\n",
       " '04_calling/QC/proband.eval.grp',\n",
       " '04_calling/QC/proband.eval.grp']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_final_output(wildcards)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
